packages/astro/src/core/session.ts

```diff
@@ -0,0 +1,406 @@
+				message: 'Invalid cookie name. Cookie names can only contain letters, numbers, and dashes.',
+			});
+		}
+		const cookieOptions: AstroCookieSetOptions = {
+			sameSite: 'lax',
+			secure: true,
+			path: '/',
+			...this.#cookieConfig,
+			httpOnly: true,
+		};
+		const value = this.#ensureSessionID();
+		this.#cookies.set(this.#cookieName, value, cookieOptions);
+	}
+
+	/**
+	 * Attempts to load the session data from storage, or creates a new data object if none exists.
+	 * If there is existing sparse data, it will be merged into the new data object.
+	 */
+
+	async #ensureData() {
+		await this.#ensureStorage();
+		if (this.#data && !this.#sparse) {
+			return this.#data;
+		}
+		this.#data ??= new Map();
+
+		// We stored this as a devalue string, but unstorage will have parsed it as JSON
+		const raw = await this.#storage?.get<any[]>(this.#ensureSessionID());
+		if (!raw) {
+			return this.#data;
+		}
+
+		try {
+			const storedMap = unflatten(raw, {
+				// Revive URL objects
+				URL: (href) => new URL(href),
+			});
+			if (!(storedMap instanceof Map)) {
+				await this.#destroySafe();
+				throw new AstroError({
+					...SessionStorageInitError,
+					message: SessionStorageInitError.message(
+						'The session data was an invalid type.',
+						this.#config.driver,
+					),
+				});
+			}
+
+			if (this.#sparse) {
+				// For sparse updates, only copy values from storage that:
+				// 1. Don't exist in memory (preserving in-memory changes)
+				// 2. Haven't been marked for deletion
+				for (const [key, value] of storedMap) {
+					if (!this.#data.has(key) && !this.#toDelete.has(key)) {
+						this.#data.set(key, value);
+					}
+				}
+			} else {
+				this.#data = storedMap;
+			}
+
+			this.#sparse = false;
+			return this.#data;
+		} catch (err) {
+			await this.#destroySafe();
+			if (err instanceof AstroError) {
+				throw err;
+			}
+			throw new AstroError(
+				{
+					...SessionStorageInitError,
+					message: SessionStorageInitError.message(
+						'The session data could not be parsed.',
+						this.#config.driver,
+					),
+				},
+				{ cause: err },
+			);
+		}
+	}
+	/**
+	 * Safely destroys the session, clearing the cookie and storage if it exists.
+	 */
+	#destroySafe() {
+		if (this.#sessionID) {
+			this.#toDestroy.add(this.#sessionID);
+		}
+		if (this.#cookieName) {
+			this.#cookies.delete(this.#cookieName);
+		}
+		this.#sessionID = undefined;
+		this.#data = undefined;
+		this.#dirty = true;
+	}
+
+	/**
+	 * Returns the session ID, generating a new one if it does not exist.
+	 */
+	#ensureSessionID() {
+		this.#sessionID ??= this.#cookies.get(this.#cookieName)?.value ?? crypto.randomUUID();
+		return this.#sessionID;
+	}
+
+	/**
+	 * Ensures the storage is initialized.
+	 * This is called automatically when a storage operation is needed.
+	 */
+	async #ensureStorage() {
+		if (this.#storage) {
+			return;
+		}
+		// Hacky way to mock storage driver for tests
+		if (
+			process.env.NODE_TEST_CONTEXT &&
```

Comment:
Is there a link to some Node.js docs about this environment variable? I would add it here for the future